src/include/dns-server/dialog-main.rb

```diff
@@ -1876,38 +1870,72 @@ def HandleExpertZonesPage(key, event)
     end
 
     # Write settings dialog
-    # @return `abort if aborted and `next otherwise
+    #
+    # @return [Symbol] :next whether configuration is saved successfully
+    #                  :back if user decided to change settings
+    #                  :abort otherwise
     def WriteDialog
-      Wizard.RestoreHelp(Ops.get_string(@HELPS, "write", ""))
-      ret = DnsServer.Write
-      if ret
-        service.reload if service.running? && status_widget.reload_flag?
-        :next
-      else
-        if Popup.YesNo(_("Saving the configuration failed. Change the settings?"))
-          :back
-        else
-          :abort
-        end
-      end
+      log.info("Running write dialog")
+
+      Wizard.RestoreHelp(write_help_text)
+
+      return :next if write_settings
+      return :back if back_to_change_setting?
+      :abort
     end
 
-    # Writes settings and restores the dialog without exiting
-    def SaveAndRestart
+    # Writes settings without exiting
+    def SaveAndRestart(event)
+      return nil unless validate_and_save_widgets(event)
+
       Wizard.CreateDialog
-      Wizard.RestoreHelp(Ops.get_string(@HELPS, "write", ""))
-      ret = DnsServer.Write
-      if ret
-        service.reload if service.running? && status_widget.reload_flag?
-      else
-        Report.Error(_("Saving the configuration failed"))
-      end
-      Builtins.sleep(1000)
-      UI.CloseDialog
+      Wizard.RestoreHelp(write_help_text)
+      Report.Error(_("Saving the configuration failed")) unless write_settings
+      Wizard.CloseDialog
+
+      service_widget.refresh
 
       nil
     end
 
+    # Writes DNS server settings and save the service
+    #
+    # @note currently, the DnsServer is a Perl module, reason why the write of
+    # settings is being performed in two steps.
+    #
+    # @return [Boolean] true if settings are saved successfully; false otherwise
+    def write_settings
+      DnsServer.Write && service.save
+    end
+
+    # Shows a popup asking to user if wants to change settings
+    #
+    # @return [Boolean] true if user decides to go back to change settings; false otherwise
+    def back_to_change_setting?
+      change_settings_message = _("Saving the configuration failed. Change the settings?")
```

Comment:
We are keeping the original text (which is right). But it does not sound good to me. I would go for something like "Do you want to review the settings?". But, for the time being, let's keep it as it is because it is out of the scope of this PR.

Reply:
I agree